controller/grpc.go

```diff
@@ -38,157 +26,25 @@ import (
-		shutdown.Fatalf("failed to create listener: %v", err)
-	}
-	logger.Debug("listener created")
-	shutdown.BeforeExit(func() { l.Close() })
-	runServer(s, config, l)
-	logger.Debug("servers stopped")
-}
-
-func runServer(s *grpc.Server, c *Config, l net.Listener) {
-	logger.Debug("loading JSON schemas...")
-
-	if err := schema.Load(schemaRoot); err != nil {
-		shutdown.Fatal(err)
-	}
-
-	logger.Debug("initializing grpc-web server...")
-	grpcWebServer := grpcweb.WrapServer(s)
-
-	logger.Debug("initializing cmux listeners...")
-	m := cmux.New(l)
-	grpcListener := m.MatchWithWriters(cmux.HTTP2MatchHeaderFieldSendSettings("content-type", "application/grpc"))
-	grpcWebListener := m.Match(cmux.Any())
-
-	statusHandler := flynnstatus.Handler(func() flynnstatus.Status {
-		if err := c.DB.Exec("ping"); err != nil {
-			return flynnstatus.Unhealthy
-		}
-		return flynnstatus.Healthy
-	})
-
-	var wg sync.WaitGroup
-
-	logger.Debug("starting servers...")
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		logger.Debug("starting gRPC server...")
-		s.Serve(grpcListener)
-	}()
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		logger.Debug("starting gRPC-web server...")
-		http.Serve(
-			grpcWebListener,
-			httphelper.ContextInjector(
-				"controller/grpc [gRPC-web]",
-				httphelper.NewRequestLogger(corsHandler(http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-					if req.URL.Path == flynnstatus.Path {
-						statusHandler.ServeHTTP(w, req)
-						return
-					}
-
-					grpcWebServer.ServeHTTP(w, req)
-				}))),
-			),
-		)
-	}()
+type grpcAPI struct {
+	*controllerAPI
 
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		logger.Debug("starting mux server...")
-		m.Serve()
-	}()
-
-	wg.Wait()
-}
-
-type Config struct {
-	logger           log.Logger
-	DB               *postgres.DB
-	q                *que.Client
-	appRepo          *data.AppRepo
-	artifactRepo     *data.ArtifactRepo
-	releaseRepo      *data.ReleaseRepo
-	formationRepo    *data.FormationRepo
-	deploymentRepo   *data.DeploymentRepo
-	eventRepo        *data.EventRepo
-	eventListenerMtx sync.Mutex
-	eventListener    *data.EventListener
-	authorizer       *utils.Authorizer
-}
-
-func configureRepos(c *Config) *Config {
-	c.appRepo = data.NewAppRepo(c.DB, os.Getenv("DEFAULT_ROUTE_DOMAIN"), routerc.New())
-	c.artifactRepo = data.NewArtifactRepo(c.DB)
-	c.releaseRepo = data.NewReleaseRepo(c.DB, c.artifactRepo, c.q)
-	c.formationRepo = data.NewFormationRepo(c.DB, c.appRepo, c.releaseRepo, c.artifactRepo)
-	c.eventRepo = data.NewEventRepo(c.DB)
-	c.deploymentRepo = data.NewDeploymentRepo(c.DB, c.appRepo, c.releaseRepo, c.formationRepo)
-	return c
+	db *postgres.DB
 }
 
 const ctxKeyFlynnAuthKeyID = "flynn-auth-key-id"
 
-func (c *Config) Authorize(ctx context.Context) (context.Context, error) {
+func (g *grpcAPI) authorize(ctx context.Context) (context.Context, error) {
 	if md, ok := metadata.FromIncomingContext(ctx); ok {
 		if passwords, ok := md["auth-key"]; ok && len(passwords) > 0 {
```

Comment:
grpc-web requests are also being checked against the `Authorization` header in the main controller. We should probably be using that here too instead of `Auth-Key` to avoid needing to use both for grpc-web requests.

Reply:
I've moved the grpcweb handler to before the basic auth check in 00463654, meaning grpcweb requests only need to specify the `Auth-Key` header.